edalize/symbiflow.py

```diff
@@ -0,0 +1,176 @@
+import platform
+import re
+import subprocess
+
+from edalize.edatool import Edatool
+from edalize.yosys import Yosys
+from importlib import import_module
+
+logger = logging.getLogger(__name__)
+
+""" Symbiflow backtend
+
+A core (usually the system core) can add the following files:
+
+- Standard design sources (Verilog only)
+
+- Constraints: unmanaged constraints with file_type SDC, pin_constraints with file_type PCF and placement constraints with file_type xdc
+
+"""
+
+class Symbiflow(Edatool):
+
+    argtypes = ['vlogdefine', 'vlogparam', 'generic']
+
+    @classmethod
+    def get_doc(cls, api_ver):
+        if api_ver == 0:
+            symbiflow_help = {
+                    'members' : [
+                        {'name' : 'package',
+                         'type' : 'String',
+                         'desc' : 'FPGA chip package (e.g. clg400-1)'},
+                        {'name' : 'part',
+                         'type' : 'String',
+                         'desc' : 'FPGA part type (e.g. xc7a50t)'},
+                        {'name' : 'builddir',
+                         'type' : 'String',
+                         'desc' : 'directory where all the intermediate files will be stored (default "build")'},
+                        {'name' : 'vendor',
+                         'type' : 'String',
+                         'desc' : 'Target architecture. Currently only "xilinx" is supported'},
+                        {'name' : 'pnr',
+                         'type' : 'String',
+                         'desc' : 'Place and Route tool. Currently only "vpr" and "nextpnr" are supported'},
+                   ]}
+
+            symbiflow_members = symbiflow_help['members']
+
+            return {'description' : "The Symbilow backend executes Yosys sythesis tool and VPR place and route. It can target multiple different FPGA vendors",
+                    'members': symbiflow_members}
+
+    def get_version(self):
+        return "1.0"
+
+    def configure_nextpnr(self):
+        (src_files, incdirs) = self._get_fileset_files(force_slash=True)
+
+        nextpnr_edam = {
+                'files'         : self.files,
+                'name'          : self.name,
+                'toplevel'      : self.toplevel,
+                'tool_options'  : {'nextpnr' : {
+                                        'arch' : 'xilinx',
+                                        'nextpnr_as_subtool' : True,
+                                        }
+
+                                }
+                }
+
+        nextpnr = getattr(import_module("edalize.nextpnr"), 'Nextpnr')(nextpnr_edam, self.work_root)
+        nextpnr.configure()
+
+        builddir = self.tool_options.get('builddir', 'build')
+
+        part = self.tool_options.get('part', None)
+        package = self.tool_options.get('package', None)
+
+        assert part is not None, 'Missing required "part" parameter'
+        assert package is not None, 'Missing required "package" parameter'
```

Comment:
I would use `logger` instead of assertions for checking user input.

Reply:
Done